File: scripts/doc_coverage.py
#!/usr/bin/env python3
"""Measure rustdoc coverage for pdftract-core public API."""

import os
import re
from pathlib import Path
from collections import defaultdict
from typing import Dict, List, Tuple

RUST_KEYWORDS = {
    'where', 'let', 'mut', 'if', 'else', 'for', 'while', 'loop', 'match',
    'return', 'break', 'continue', 'impl', 'struct', 'enum', 'trait',
    'type', 'fn', 'const', 'static', 'mod', 'use', 'crate', 'super',
    'self', 'Self', 'extern', 'unsafe', 'async', 'await', 'move',
    'ref', 'True', 'False', 'Some', 'None', 'Ok', 'Err', 'Vec',
    'String', 'Box', 'Result', 'Option', 'u8', 'u16', 'u32', 'u64',
    'i8', 'i16', 'i32', 'i64', 'f32', 'f64', 'bool', 'usize', 'isize'
}


def extract_items_from_file(filepath: Path) -> List[Tuple[str, str, int, bool]]:
    """Extract public items from a Rust source file.

    Returns: List of (name, kind, line_number, has_example) tuples.
    """
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()

    items = []
    lines = content.split('\n')

    # Track current doc comment for next item
    pending_doc = None

    for i, line in enumerate(lines, 1):
        stripped = line.strip()

        # Skip empty lines and non-doc comments
        if not stripped or stripped.startswith('//') and not stripped.startswith('///'):
            if stripped.startswith('//') and not stripped.startswith('///'):
                pending_doc = None
            continue

        # Track doc comments
        if stripped.startswith('///'):
            if pending_doc is None:
                pending_doc = []
            pending_doc.append(stripped)
            continue

        # Check for attribute lines (cfg, derive, etc.) - don't reset doc
        if stripped.startswith('#['):
            continue

        # Check for pub items
        if stripped.startswith('pub '):
            # Extract item kind and name
            kind_match = re.search(r'pub (fn|struct|enum|trait|type|const|mod|use)\s+(\w+)', stripped)
            if not kind_match:
                # Handle complex cases like `pub use foo::Bar;`
                use_match = re.search(r'pub use\s+(.+?);', stripped)
                if use_match:
                    item_name = use_match.group(1).split('::')[-1].rstrip(';')
                    kind = 'use'
                else:
                    continue
            else:
                kind = kind_match.group(1)
                item_name = kind_match.group(2)

            # Skip known items that are re-exports
            if item_name in RUST_KEYWORDS:
                pending_doc = None
                continue

            # Check if doc has examples
            has_example = False
            if pending_doc:
                doc_text = '\n'.join(pending_doc)
                has_example = '```rust' in doc_text or '```no_run' in doc_text

            items.append((item_name, kind, i, has_example))
            pending_doc = None

        # Reset doc if we encounter something else
        elif stripped and not stripped.startswith('#') and not stripped.startswith('use'):
            pending_doc = None

    return items


def scan_directory(src_dir: Path) -> Dict[str, List[Tuple[str, str, int, bool]]]:
    """Scan all Rust files in a directory."""
    all_items = {}

    for rust_file in src_dir.rglob('*.rs'):
        # Skip test files and tests modules
        if 'tests.rs' in rust_file.name or 'test_' in rust_file.name:
            continue
        if any(p.startswith('test') or p == 'benches' for p in rust_file.parts):
            continue

        relative = rust_file.relative_to(src_dir)
        module_path = str(relative.with_suffix(''))

        items = extract_items_from_file(rust_file)
        if items:
            all_items[module_path] = items

    return all_items


def print_report(all_items: Dict[str, List[Tuple[str, str, int, bool]]]):
    """Print coverage report."""
    total = 0
    with_examples = 0
    by_kind = defaultdict(lambda: [0, 0])  # kind -> [total, with_examples]

    print("=" * 80)
    print("RUSTDOC COVERAGE REPORT")
    print("=" * 80)

    for module_path in sorted(all_items.keys()):
        items = all_items[module_path]
        if not items:
            continue

        module_total = len(items)
        module_with = sum(1 for _, _, _, has_ex in items if has_ex)
        module_pct = (module_with / module_total * 100) if module_total else 0

        print(f"\n{module_path}:")
        print(f"  {module_with}/{module_total} items with examples ({module_pct:.1f}%)")

        # List missing examples
        missing = [name for name, kind, _, has_ex in items if not has_ex and kind in ('fn', 'struct', 'enum', 'trait', 'type')]
        if missing:
            print(f"  Missing examples: {', '.join(missing[:10])}", end='')
            if len(missing) > 10:
                print(f" ... and {len(missing) - 10} more")
            else:
                print()

        total += module_total
        with_examples += module_with

        for _, kind, _, has_ex in items:
            by_kind[kind][0] += 1
            if has_ex:
                by_kind[kind][1] += 1

    overall_pct = (with_examples / total * 100) if total else 0
    print("\n" + "=" * 80)
    print(f"OVERALL: {with_examples}/{total} items with examples ({overall_pct:.1f}%)")
    print("=" * 80)

    print("\nBy kind:")
    for kind in sorted(by_kind.keys()):
        t, w = by_kind[kind]
        pct = (w / t * 100) if t else 0
        print(f"  {kind:10s}: {w:4d}/{t:4d} ({pct:5.1f}%)")

    # Threshold check
    print("\n" + "=" * 80)
    if overall_pct >= 80:
        print("PASS: Meets 80% threshold")
    else:
        print(f"FAIL: Below 80% threshold (need {int((0.8 * total) - with_examples)} more examples)")
    print("=" * 80)


if __name__ == '__main__':
    src_dir = Path('/home/coding/pdftract/crates/pdftract-core/src')
    all_items = scan_directory(src_dir)
    print_report(all_items)

File: scripts/doc_coverage.rs
#!/usr/bin/env python3
"""Count public items in pdftract-core and measure documentation coverage."""

import subprocess
import json
import re
from pathlib import Path
from typing import Dict, List, Tuple

def run_cargo_doc() -> str:
    """Run cargo doc and capture output."""
    result = subprocess.run(
        ["cargo", "doc", "--no-deps", "--all-features", "-p", "pdftract-core"],
        cwd=Path("/home/coding/pdftract"),
        capture_output=True,
        text=True
    )
    return result.stdout + result.stderr

def has_example(doc: str) -> bool:
    """Check if documentation contains a code example."""
    if not doc:
        return False
    # Look for ```rust, ```no_run, ```ignore, etc.
    return bool(re.search(r'```rust', doc))

def extract_docs_from_file(file_path: Path) -> List[Tuple[str, str, bool, str]]:
    """Extract public items and their docs from a Rust file."""
    items = []

    content = file_path.read_text()
    lines = content.split('\n')

    # Track current doc comment being built
    current_doc = []
    doc_line_start = 0

    for i, line in enumerate(lines):
        stripped = line.strip()

        # Check for doc comments
        if stripped.startswith("///"):
            current_doc.append(stripped[3:].strip())
            if not doc_line_start:
                doc_line_start = i + 1
        elif stripped.startswith("//!"):
            # Module-level doc - skip for item-level tracking
            pass
        elif stripped.startswith("//"):
            # Regular comment - skip
            pass
        else:
            # Check if this is a public item declaration
            if current_doc:
                pub_match = re.match(r'pub\b\s*(fn|struct|enum|trait|type|const|static|mod)\b\s*(\w+)?', stripped)
                if pub_match:
                    item_type = pub_match.group(1)
                    item_name = pub_match.group(2) or f"anon_{i}"
                    doc_text = "\n".join(current_doc)
                    items.append((item_type, item_name, has_example(doc_text), file_path.name))
                current_doc = []
                doc_line_start = 0

    return items

def main():
    """Main entry point."""
    print("Checking pdftract-core documentation coverage...\n")

    # First, run cargo doc to check for warnings
    print("Running cargo doc --no-deps --all-features...")
    result = subprocess.run(
        ["cargo", "doc", "--no-deps", "--all-features", "-p", "pdftract-core"],
        cwd=Path("/home/coding/pdftract"),
        capture_output=True,
        text=True
    )

    has_warnings = "warning:" in result.stdout or "warning:" in result.stderr
    has_missing_docs = "missing documentation" in result.stdout or "missing documentation" in result.stderr

    if has_warnings:
        print("⚠️  Warnings found:")
        for line in (result.stdout + result.stderr).split('\n'):
            if 'warning:' in line or 'warning:' in line.lower():
                print(f"  {line.strip()}")
    elif has_missing_docs:
        print("❌ Missing documentation warnings found")
    else:
        print("✅ No warnings - cargo doc passes!")

    print("\nScanning source files for public items with examples...")

    src_dir = Path("/home/coding/pdftract/crates/pdftract-core/src")
    all_items: List[Tuple[str, str, bool, str]] = []

    for rs_file in src_dir.rglob("*.rs"):
        if rs_file.name == "lib.rs":
            continue  # Already well-documented
        items = extract_docs_from_file(rs_file)
        all_items.extend(items)

    # Count by category
    total_items = len(all_items)
    items_with_examples = sum(1 for _, _, has_ex, _ in all_items if has_ex)
    coverage = (items_with_examples / total_items * 100) if total_items > 0 else 0

    print(f"\n📊 Documentation Coverage:")
    print(f"  Total public items: {total_items}")
    print(f"  With examples: {items_with_examples}")
    print(f"  Coverage: {coverage:.1f}%")

    # Show items without examples by type
    by_type: Dict[str, List[Tuple[str, bool, str]]] = {}
    for item_type, item_name, has_ex, file_name in all_items:
        if item_type not in by_type:
            by_type[item_type] = []
        by_type[item_type].append((item_name, has_ex, file_name))

    print(f"\n📋 By item type:")
    for item_type, items in sorted(by_type.items()):
        with_ex = sum(1 for _, h, _ in items if h)
        total = len(items)
        cov = (with_ex / total * 100) if total > 0 else 0
        print(f"  {item_type}: {with_ex}/{total} ({cov:.0f}%)")

    # Find high-value modules needing examples
    print(f"\n🔍 High-value modules needing examples:")
    high_value_modules = [
        "extract.rs", "document.rs", "parser/mod.rs", "span/mod.rs",
        "table/mod.rs", "layout/mod.rs", "output/mod.rs"
    ]
    for mod_name in high_value_modules:
        mod_items = [(t, n, h) for t, n, h, f in all_items if f == mod_name]
        if mod_items:
            with_ex = sum(1 for _, _, h in mod_items if h)
            total = len(mod_items)
            cov = (with_ex / total * 100) if total > 0 else 0
            if cov < 80:
                print(f"  {mod_name}: {with_ex}/{total} ({cov:.0f}%)")

    # Check against threshold
    if coverage >= 80:
        print(f"\n✅ PASS: {coverage:.1f}% >= 80% threshold")
        return 0
    else:
        print(f"\n❌ FAIL: {coverage:.1f}% < 80% threshold")
        print(f"   Need {int((80 - coverage) / 100 * total_items)} more items with examples")
        return 1

if __name__ == "__main__":
    exit(main())
